CR-TEST: Crunch DB Fitness (Slobodan Ilic)

slobodan-ilic opened the pull request:
This PR represents the work that has been done on the assigned challenge (cr-db). The PR is opened as per request of @percious. All comments, suggestions and critiques are welcome.

slobodan-ilic committed:
Prepare working environment for Crunch test

1. Ignore *.swp files
2. Write a JSON generating script (for 1st task)
3. Fix headers of the 10k CSV file (for 2nd task)
4. Edit cr-db/setup.py to include dependencies

slobodan-ilic committed:
1st Task - Optimize insertion into Mongo

Implement two types of optimization
for speed and memory

slobodan-ilic committed:
2nd Task - Modify raw CSV data loading into MongoDB

Create two type of modifications
- convert and store categoty data to integers
- convert category data to integers and
  store as joined strings
- add unit test for column modification
- add integration test for columns comparison

slobodan-ilic committed:
3rd Task - Implement test with filtering data

Extract data for average adjusted salary, based on
formal education, for women. Plot the result as bar-chart.

slobodan-ilic committed:
Update readme file

slobodan-ilic committed:
Encapsulate functionality into a Loader class

- Create a loader class and add all relevant functions to it
- Extract the functionality from the filtering test, and
  place it into a function in the Loader class

percious commented on cr-db/cr/db/loader.py:
```diff
@@ -1,53 +1,158 @@
-        headers = csv_data.next()
+        with file(filename) as the_file:
+            if optimize_for == 'speed':
+                objs = json.load(the_file)
```
did you find ujson?  I wonder how this might affect your benchmarks.

slobodan-ilic replied:
@percious Did the tests with ujson, notable performance improvement on data set with 1M users. Added comments in the commit message.

percious commented on cr-db/cr/db/loader.py:
```diff
@@ -1,53 +1,158 @@
+                    headers[i] = last_header
+
+            columns = [[] for _ in headers]
+            for _, row in enumerate(csv_data):
```
why enumerate if you aren't using the index?

slobodan-ilic replied:
@percious The `enumerate` was there before I did the challenge (and the index `r` wasn't used). I saw that `r` was not needed, and replaced it with a `_`, but failed to realize that I could remove the `enumerate` as well. I'll amend the commit.

slobodan-ilic replied:
@percious Removed unnecessary enumerate

percious commented on cr-db/cr/db/loader.py:
```diff
@@ -1,53 +1,158 @@
+            return db.datasets.insert(data, check_keys=False)
+
+    @classmethod
+    def select_filtered_data(cls, dataset, filter_col, filter_val, group_by,
```
why did you choose to persist the filtering rather than allow the filters to happen "on the fly".

slobodan-ilic replied:
@percious I'm not sure what you mean by 'persist'. The filter is passed in as a function parameter, which allows it to take multiple values in different scenarios. In order to generate the requested data, we have the following steps:
1. Selection
2. Filtering
3. Grouping
4. Preprocessing
And they need to be in that particular order, since some of the groups may be filtered out in the 2nd step.

Anyhow, I've factored out the category filtering, for a particular column, into a separate function. This allows for better reusability. Also, the preparation of the data is factored out into a separate function (packaging into numpy array).

Please let me know if you had something else in mind...

percious commented on cr-db/cr/db/loader.py:
```diff
@@ -1,53 +1,158 @@
+        with file(filename) as the_file:
+            if optimize_for == 'speed':
+                objs = json.load(the_file)
+                collection.insert(objs)
```
also, if you know you are dealing with a file, it might make sense to skip the encode/decode altogether and load directly into mongo.

slobodan-ilic replied:
@percious Did an example/test with GridFS, and directly storing a file into MongoDB.

percious commented on cr-db/tests/test_db.py:
```diff
@@ -5,47 +5,116 @@
+    # Even with converting everything to int, bool, and None, I was only able
+    # to get the BSON document size from about 66M down to about 30M, and the
+    # max size set in MongoDB is about 16M. So I resorted to storing the data
+    # as joined strings, which worked just fine.
```
why used joined strings?  If you know there are only 2^4 categories in the given column, could you not use UInt in the BSON to get the size down?  I think the idea here is to _not_ store the data as strings.  Also, I'm not entirely clear, are the categories you are creating encoded automatically by finding all of the possible categories and can you point me to the line that does this?

slobodan-ilic replied:
@percious The joined strings are a conversion option that allowed me to complete the task.The size of BSON with integers just couldn't get reduced more than about 30M on my machine. I tried a number of different python types, including byte array and raw binary BSON, but they didn't help. I also tried using `None`s instead of `0`s, but that only helped insignificantly.

This is a kind of a problem for which I'd immediately ask a senior person (about the UInt in BSON, or the similar guideline). Especially so, since I didn't have prior experience with BSON or Mongo.

As for all possible categories encoding, it's done in [this](https://github.com/percious/crunch-fitness/pull/3/files#diff-0b0c428d576620398241ea0d50a1b6e7R135) part of the code. The `set` function does the reduction to unique values. Only the columns where `max_categories` number of categories is not exceeded are actually converted, so that we retain  numerical values (such as the salary).

I'll do the changes to BSON, to use the UInt, and amend the commit. Thanks for the info :100:

slobodan-ilic replied:
@percious I'm looking at how to use the UInt for BSON, but I'm not able to find that type in [bson types](https://docs.mongodb.com/v3.0/reference/bson-types/), nor in pymongo. Could you please point me to an example (or documentation) that you had in mind? I'm running into same snippets of documentation that I was running into when I originally researched this, but still no luck...

Also, about your comment about storing data to strings (and that not being the point)... Please note that the data is converted to a numerical category representation, prior to joining into a string. There is actual compression happening.

percious reviewed (commented):
Some feedback.

slobodan-ilic commented:
@percious I've addressed some of the comments, I plan on addressing others as well. Will let you know when I do. I'll commit the changes in separate commits, so that you can easily detect them. I can later squash them if needed.

slobodan-ilic committed:
Address PR review comment No 1

- Removed unnecessary enumeration
- Tested performance with ujson, clear visible improvement, about 20-30%
- Added loading to file to MongoDB
- Optimized for memory usage and speed, allowed grouped inserts
  up to a certain number of elements per insert

slobodan-ilic committed:
Address review comment No 2

- Modify columns to integers only
- Add encoding functions for storing in MongoDB as bytearray
- Fix existing tests and add more unittests

slobodan-ilic committed:
Address review comment No 3

- Factor out category filtering into a separate function
- Factor out data preparation into a separate function

slobodan-ilic commented:
@percious I've addressed all the comments.
